src/mono/mono/mini/aot-compiler.c

```diff
@@ -5243,9 +5243,16 @@ MONO_RESTORE_WARNING
 					if (decoded_args->named_args_info [j].field && !strcmp (decoded_args->named_args_info [j].field->name, "EntryPoint")) {
 						named = (const char *)decoded_args->named_args[j]->value.primitive;
 						slen = mono_metadata_decode_value (named, &named);
+#ifdef TARGET_MACH
```

Comment:
use acfg->user_symbol_prefix instead it will have the right prefix for exported symbols and no need to use ifdefs here.

Reply:
If you change the export_name to include the prefix, then you need to adjust g_string_append_printf (export_symbols, "%s%s\n", acfg->user_symbol_prefix, export_name);

Reply:
Thank you, I have corrected the code based on your suggestions. 
Should we also cover handling of `MonoPInvokeCallbackAttribute` (I am not familiar of how exported symbols should be mangled in that case) ?

Reply:
That is pretty old logic and its not documented as part of MonoPInvokeCallbackAttribute and I have personally not used it to export out native->managed wrappers. Looking through aot-compiler.c all other exports added to acfg->export_names now uses acfg->user_symbol_prefix, so for consistency, I believe we should do the same for this attribute.

Reply:
This might break existing code. Are we going to use MonoPInvokeCallbackAttribute in the future ?

Reply:
I am not familiar with how the `MonoPInvokeCallbackAttribute` supposed to work, but I was wondering if the existing native code is using this feature, then how did it work so far (if we consider that `dlsym`, for example, requires the exported symbols to be prefixed with underscore on Apple platforms)?

Reply:
@lateralusX what do you think?

Reply:
Not sure, `MonoPInvokeCallbackAttribute` has been around for 11 years, https://github.com/mono/mono/commit/928bd3ee2bc77f2fbd3e4fc1a21bc1ac68df0cfe, so even if `ExportSymbol` is not documented there is a risk that it has been used with a prefix on Apple platforms. It would also work in the static linking scenario where you, so if there is code using that, it will be broken as well. We have the same risk with `UnmanagedCallersOnlyAttribute` as well, since it has been released and used in code and might break as well with this change. One alternative could be to export the symbols twice on platforms that uses a prefix, to maintain backward compatibility and since the number of functions using these attributes and export symbol names should be fairly small and it only applies to some platforms, I don't think that would have to big of an impact. @vargaz do you think it would be an issue if we exported them using both old name as well as name with prefix on platforms that uses it?

Reply:
If we take the approach of emitting both symbols, should both be added to the `exports file`?

Reply:
No just the correct one, exports file is a new feature, so no backward combability issues with that one.

Reply:
I think since MonoPInvokeCallbackAttribute is not going to be used too much in the future, we shouldn't change its behavior, nobody reported a problem with the current one.

Reply:
I can revert the change regarding `MonoPInvokeCallbackAttribute` and add a warning to the compiler instead, noting that the exported symbol is not prefixed in its case?

---

>We have the same risk with UnmanagedCallersOnlyAttribute as well 

Should we mark the PR with breaking change label then?

Reply:
If decide to leave `MonoPInvokeCallbackAttribute` as is I don't think we need to do anything with it at all, its old, people lived with it for 11+ years and figured out how to get it to work in the process. If we should do anything then it would be to document that its deprecated in favor of `UnmanagedCallersOnlyAttribute`.

For the `UnmanagedCallersOnlyAttribute`, if we take a step back and look at the use cases from and end user perspective:

The user uses `UnmanagedCallersOnlyAttribute` to get a wrapper that can be called from native code, so it can be passed over to native code using pinvoke. There is also a way to give it a symbol name, meaning native code could get hold of the symbol and call it without need to go through managed code.

Lets assume:

```
[UnmanagedCallersOnly(EntryPoint = "MyTest")]
public static void MyTest()
{
...
}
```

A user could consume this entry point in two ways:

1. User compile the assembly into and obj file using -static aot argument, the user would then assume that he can use the symbol in an extern declare from a C source that is then statically linked with the obj. I assume this case should use the symbols name as is since its not exported from a shared library, just linked as a regular symbol and use as is from a C source file. User has something like this in a C source:

```
extern void MyTest (void);
...
MyTest ();
```

2. User compile the assembly into shared library (or into an obj that user intend to include in a shared library). In this case consumer of symbol will use dlsym to load the symbol using the given name, meaning it needs to use any platform prefix needed. User has something like this in a C source:

```
typedef void(*MyTestFunc)(void);
MyTestFunc MyTest = (MyTestFunc)dlsym(handle, "MyTest");
...
MyTest ();
```

So in order for us to support both scenarios, I believe we need to emit symbols supporting both use cases, meaning that we would need to emit two symbols on platform that needs prefix for symbols exported out from a shared library.

Reply:
@lateralusX regarding your comment `1.` 
- If I create a `native.c`, add the example code you suggested, and build it with:
`clang -c native.c -o native.o`
- and inspect symbols with:
`nm native.o > syms.list`
- it will show the extern symbol with the underscore:
```
                 U _MyTest
0000000000000000 T _main
0000000000000000 t ltmp0
0000000000000030 s ltmp1
```
Which means that in both cases it is expected to have the underscore prefix. Am I possibly missing something?

---
Info about the toolchain used:
```
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.2.0
Thread model: posix
InstalledDir: /Applications/Xcode_14.1.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Reply:
You need to have our aot compiler create the symbol and then try to link using that object file to see if you even can link the static scenario using extern declared C function. It might not even work if toolchain will always decorate with _ even for the static scenario not just when exporting out of a shared library.

Reply:
> You need to have our aot compiler create the symbol (since it will emit it without prefix) and then try to link using that object file to see if you even can link the static scenario using extern declared C function. It might not even work if toolchain will always decorate with _ even for the static scenario not just when exporting out of a shared library.

I tried the following: 
If I revert my change of decorating exported symbols with a prefix, and adapt the `main.m`, of the added functional test, to reference `exposed_managed_method` through a extern declared C function (instead of dynamically looking up the symbol) the build of the test fails with: 
```
Undefined symbols for architecture arm64:
  "_exposed_managed_method", referenced from:
      ___29-[ViewController viewDidLoad]_block_invoke in main.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
as the linker seems to expect that in the static scenario it also expects the symbol to be prefixed with the underscore. With my change the error goes away, the app can be built, run and the exposed managed method is properly invoked.

Reply:
Thanks for looking into it. So then the symbol needs to be prefixed for both scenario (static, shared), that also means if any existing code would have this work in any scenario, then the user needed to manually have prefixed the symbol on Apple platforms.

Looks like this, https://github.com/lambdageek/dnne_monovm_sample/tree/main, does use UnmanagedCallersOnly with EntryPoint on Mono, curious how that scenarios works, since it sounds like it should have hit issues with unresolved symbols and I know the author is on OSX :), possible to take a quick look to see if that scenario gets broken with this change?

Reply:
> I assume that linking a shared library directly against application also expects symbols being exported with prefixed entrypoints?

I have tested this by:
1. Creating a HelloWorld console application which exposes:
``` csharp
[UnmanagedCallersOnly(EntryPoint="exposed_managed_method")]
public static int ManagedMethod() => 42;
```
2. FullAOT compiled the application as a shared library
3. Create a simple `main.c`:
``` c
int exposed_managed_method();

int main()
{
    const char *managed_argv [128];
    managed_argv[0] = "/Users/ivan/repos/runtime-mono-AOT/artifacts/bin/HelloWorld/arm64/Debug/osx-arm64/publish/HelloWorld.dll";
    int res;
    printf("hello from native");
    mono_jit_set_aot_mode (MONO_AOT_MODE_FULL);
    setenv("MONO_PATH", "/Users/ivan/repos/runtime-mono-AOT/artifacts/bin/HelloWorld/arm64/Debug/osx-arm64/publish", 1);
    MonoDomain *domain = mono_jit_init ("HelloWorld");
    const char* executable = "HelloWorld.dll";
    MonoAssembly *assembly = mono_assembly_open (executable, NULL);
    res = mono_jit_exec (mono_domain_get (), assembly, 1, managed_argv);
    printf("Invoking exposed method: %d", exposed_managed_method());
    return 0;
}
```
4. Compiled the native code with:
```
clang -I /Users/ivan/repos/runtime-mono-AOT/artifacts/bin/microsoft.netcore.app.runtime.osx-arm64/Debug/runtimes/osx-arm64/native/include/mono-2.0 \
main.c \
-force_load /Users/ivan/repos/runtime-mono-AOT/artifacts/obj/mono/osx.arm64.Debug/cross/mono/mini/libmonosgen-2.0.a \
-force_load libSystem.Native.a \
-lz \
HelloWorld.dll.dylib \
-framework Foundation \
-g \
-o main
```
5. The code properly runs and prints the desired output: `./main`
---
Conclusion: It seems that in all cases the symbol is expected to be prefixed with underscore

Reply:
Excellent! Thanks for investigating!